text/0000-must-not-await-lint.md

```diff
@@ -0,0 +1,152 @@
+
+```rust
+#[must_not_await = "Your error message here"]
+struct MyStruct {}
+```
+
+This struct if held across an await boundary would cause a deny-by-default warning:
+
+```rust
+async fn foo() {
+  let my_struct = MyStruct {};
+  my_async_op.await;
+  println!("{:?}", my_struct);
+}
+```
+
+The compiler might output something along the lines of:
+
+```
+warning: `MyStruct` should not be held across an await point.
+```
+
+Example use cases for this lint:
+
+- `MutexGuard` holding this across a yield boundary in a single threaded executor could cause deadlocks. In a multi-threaded runtime the resulting future would become `!Send` which will stop the user from spawning this future and causing issues. But in a single threaded runtime which accepts `!Send` futures deadlocks could happen.
+
+- The same applies to other such synchronization primitives such as locks from `parking-lot`.
+
+- `tracing::Span` has the ability to enter the span via the `tracing::span::Entered` guard. While entering a span is totally normal, during an async fn the span only needs to be entered once before the `.await` call, which might potentially yield the execution.
+
+- Any RAII guard might possibly create unintended behavior if held across an await boundary.
+
+This lint will enable the compiler to warn the user that the code could produce unforeseen side effects. Some examples of this are:
+
+- [`std::sync::MutexGuard`](https://doc.rust-lang.org/std/sync/struct.MutexGuard.html)
+- [`tracing::span::Entered`](https://docs.rs/tracing/0.1.15/tracing/span/struct.Entered.html)
+
+This will be a best effort lint to signal the user about unintended side-effects of using certain types across an await boundary.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The `must_not_await` attribute is used to issue a diagnostic warning when a value is not "used". It can be applied to user-defined composite types (structs, enums and unions), functions and traits.
+
+The `must_not_await` attribute may include a message by using the [`MetaNameValueStr`] syntax such as `#[must_not_await = "example message"]`.  The message will be given alongside the warning.
+
+When used on a user-defined composite type, if a value exists across an await point, then this lint is violated.
+
+
+```rust
+#[must_not_await = "Your error message here"]
+struct MyStruct {}
+
+async fn foo() {
+  let my_struct = MyStruct {};
+  my_async_op.await;
+  println!("{:?}", my_struct);
+}
+```
+
+When used on a function, if the value returned by a function is held across an await point, this lint is violated.
+
+```rust
+#[must_not_await]
+fn foo() -> i32 { 5i32 }
+
+async fn foo() {
+  let bar = foo();
+  my_async_op.await;
+  println!("{:?}", bar);
+}
+```
+
+When used on a [trait declaration], if the value implementing that trait is held across an await point, the lint is violated.
```

Comment:
This could use some terminology clarification, to disambiguate between attributes on the `trait` itself and attributes on a `fn` within the `trait`.